go/vt/vtgate/planbuilder/operators/union_merging.go

```diff
@@ -108,40 +109,70 @@ func mergeUnionInputsInOrder(ctx *plancontext.PlanningContext, op *Union) ([]ops
 // If they can be merged, a new operator with the merged routing is returned
 // If they cannot be merged, nil is returned.
 // this function is very similar to mergeJoinInputs
-func mergeUnionInputs(ctx *plancontext.PlanningContext, lhs, rhs ops.Operator, lhsExprs, rhsExprs sqlparser.SelectExprs, distinct bool) (ops.Operator, sqlparser.SelectExprs, error) {
+func mergeUnionInputs(
+	ctx *plancontext.PlanningContext,
+	lhs, rhs ops.Operator,
+	lhsExprs, rhsExprs sqlparser.SelectExprs,
+	distinct bool,
+) (ops.Operator, sqlparser.SelectExprs, error) {
 	lhsRoute, rhsRoute, routingA, routingB, a, b, sameKeyspace := prepareInputRoutes(lhs, rhs)
 	if lhsRoute == nil {
 		return nil, nil, nil
 	}
 
 	switch {
 	// if either side is a dual query, we can always merge them together
-	case b == dual:
+	// an unsharded/reference route can be merged with anything going to that keyspace
+	case b == dual || (b == anyShard && sameKeyspace):
 		return createMergedUnion(ctx, lhsRoute, rhsRoute, lhsExprs, rhsExprs, distinct, routingA)
-	case a == dual:
+	case a == dual || (a == anyShard && sameKeyspace):
```

Comment:
nit: parenthesis around the `&&` block are not needed

Reply:
I think the parenthesis improves readability